File: plugin/python/larch/review/tally_engine.py
"""Shared, ordered per-item adjudication for review tally families."""

from __future__ import annotations

from collections.abc import Callable, Iterable
from dataclasses import dataclass, replace
from pathlib import Path

from larch.review import voting

VoteCell = tuple[str, str, str, str, str, str | None]


@dataclass(frozen=True)
class ItemContext:
    """Family-prepared inputs for one ballot item.

    Families own parsing and attribution restoration.  The engine owns only the
    policy decisions that must remain identical as tally implementations converge.
    """

    item_id: str
    block_path: Path
    block_text: str
    artifact_text: str
    reviewer: str
    cells: tuple[VoteCell, ...]
    yes: int
    no: int
    judge_error: int
    is_oos: bool
    eligible_voters: int
    voter_votes: tuple[tuple[str, str], ...]
    voter_severities: tuple[str, ...]


@dataclass(frozen=True)
class ItemAdjudicationResult:
    """Cached policy outcome, before and after the family security hook."""

    context: ItemContext
    voting_result: str
    classification_scope: str
    neutral_rescued: bool
    fileable_oos: bool
    score_kind: str
    score_result: str
    accepted_weight: int
    unique_finder_eligible: bool
    ledger_outcome: str
    reroute_marker: str
    artifact_bucket: str
    security: bool | None = None


def _finding_oos_reroute_marker(*, block_text: str, neutral_rescued: bool) -> str:
    """Return the engine-owned marker for a neutral finding rerouted to OOS."""
    _ = block_text
    return "neutral-rescued" if neutral_rescued else ""


def adjudicate_item(context: ItemContext) -> ItemAdjudicationResult:
    """Evaluate the shared policy for a fully prepared item exactly once."""
    vote_values = tuple(vote for _label, vote in context.voter_votes)
    if context.is_oos:
        voting_result = voting.classify_oos_result(
            yes=context.yes,
            no=context.no,
            exonerate=0,
            eligible=context.eligible_voters,
        )
    else:
        voting_result = voting.classify_result(
            yes=context.yes,
            no=context.no,
            exonerate=0,
            eligible=context.eligible_voters,
        )
    fileable_oos = voting.oos_fileable_from_votes(
        voting_result,
        yes_votes=vote_values,
        severities=context.voter_severities,
    )
    neutral_rescued = voting.neutral_high_severity_rescue_to_oos(
        voting_result,
        yes_votes=vote_values,
        severities=context.voter_severities,
    )
    score_kind = "oos" if context.is_oos or neutral_rescued else "finding"
    score_result = "neutral" if context.is_oos and voting_result == "accepted" and not fileable_oos else voting_result
    accepted_weight = (
        voting.accepted_finding_points_from_severities(context.voter_severities, votes=vote_values)
        if score_kind == "finding" and score_result == "accepted"
        else 0
    )
    classification_scope = "oos" if context.is_oos or neutral_rescued else "in_scope"
    ledger_outcome = "oos" if classification_scope == "oos" else voting_result
    if context.is_oos or neutral_rescued:
        artifact_bucket = "oos"
    elif voting_result == "accepted":
        artifact_bucket = "accepted"
    else:
        artifact_bucket = "rejected"
    return ItemAdjudicationResult(
        context=context,
        voting_result=voting_result,
        classification_scope=classification_scope,
        neutral_rescued=neutral_rescued,
        fileable_oos=fileable_oos,
        score_kind=score_kind,
        score_result=score_result,
        accepted_weight=accepted_weight,
        unique_finder_eligible=score_kind == "finding" and score_result == "accepted",
        ledger_outcome=ledger_outcome,
        reroute_marker=_finding_oos_reroute_marker(block_text=context.block_text, neutral_rescued=neutral_rescued),
        artifact_bucket=artifact_bucket,
    )


def run_items(
    contexts: Iterable[ItemContext],
    *,
    serialize: Callable[[ItemAdjudicationResult], None],
    security_hook: Callable[[ItemContext], bool],
    publish: Callable[[ItemAdjudicationResult], None],
) -> list[ItemAdjudicationResult]:
    """Run the required prepare → serialize → security → publish lifecycle.

    Iterating lazily is deliberate: failures leave prior output intact and do not
    prepare, serialize, or publish later ballot items.
    """
    completed: list[ItemAdjudicationResult] = []
    for context in contexts:
        result = adjudicate_item(context)
        serialize(result)
        completed_result = replace(result, security=security_hook(context))
        publish(completed_result)
        completed.append(completed_result)
    return completed

File: crates/larch-core/src/review/tally_engine.rs
//! Ordered review-item adjudication shared by tally families.

use super::voting;
use std::path::PathBuf;

/// One prepared judge cell.
pub type VoteCell = (String, String, String, String, String, Option<String>);

/// Inputs prepared by a tally family for one item.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ItemContext {
    /// Item identifier.
    pub item_id: String,
    /// Backing block path.
    pub block_path: PathBuf,
    /// Canonical block text.
    pub block_text: String,
    /// Artifact text.
    pub artifact_text: String,
    /// Proposing reviewer.
    pub reviewer: String,
    /// Parsed judge cells.
    pub cells: Vec<VoteCell>,
    /// YES count.
    pub yes: usize,
    /// NO count.
    pub no: usize,
    /// Judge-error count.
    pub judge_error: usize,
    /// Whether this started as an OOS item.
    pub is_oos: bool,
    /// Eligible voter count.
    pub eligible_voters: usize,
    /// Per-voter vote values.
    pub voter_votes: Vec<(String, String)>,
    /// Per-voter severity values.
    pub voter_severities: Vec<String>,
}

/// Policy decision before and after the security hook.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ItemAdjudicationResult {
    /// Prepared item context.
    pub context: ItemContext,
    /// Base vote classification.
    pub voting_result: String,
    /// In-scope or OOS classification.
    pub classification_scope: String,
    /// Whether major YES evidence rerouted a neutral finding.
    pub neutral_rescued: bool,
    /// Whether an OOS result is fileable.
    pub fileable_oos: bool,
    /// Score domain.
    pub score_kind: String,
    /// Score result.
    pub score_result: String,
    /// Accepted finding weight.
    pub accepted_weight: u8,
    /// Whether sole-finder bonus eligibility applies.
    pub unique_finder_eligible: bool,
    /// Ledger outcome.
    pub ledger_outcome: String,
    /// Artifact reroute marker.
    pub reroute_marker: String,
    /// Artifact bucket.
    pub artifact_bucket: String,
    /// Security classification after the security hook.
    pub security: Option<bool>,
}

/// Adjudicate a prepared item exactly once.
#[must_use]
pub fn adjudicate_item(context: ItemContext) -> ItemAdjudicationResult {
    let votes: Vec<String> = context
        .voter_votes
        .iter()
        .map(|(_, vote)| vote.clone())
        .collect();
    let voting_result = if context.is_oos {
        voting::classify_oos_result(context.yes, context.eligible_voters)
    } else {
        voting::classify_result(context.yes, context.eligible_voters)
    }
    .to_owned();
    let fileable_oos =
        voting::oos_fileable_from_votes(&voting_result, &votes, &context.voter_severities);
    let neutral_rescued = voting::neutral_high_severity_rescue_to_oos(
        &voting_result,
        &votes,
        &context.voter_severities,
    );
    let score_kind = if context.is_oos || neutral_rescued {
        "oos"
    } else {
        "finding"
    }
    .to_owned();
    let score_result = if context.is_oos && voting_result == "accepted" && !fileable_oos {
        "neutral".to_owned()
    } else {
        voting_result.clone()
    };
    let accepted_weight = if score_kind == "finding" && score_result == "accepted" {
        voting::accepted_finding_points_from_severities(&votes, &context.voter_severities)
    } else {
        0
    };
    let classification_scope = if context.is_oos || neutral_rescued {
        "oos"
    } else {
        "in_scope"
    }
    .to_owned();
    let ledger_outcome = if classification_scope == "oos" {
        "oos".to_owned()
    } else {
        voting_result.clone()
    };
    let artifact_bucket = if classification_scope == "oos" {
        "oos"
    } else if voting_result == "accepted" {
        "accepted"
    } else {
        "rejected"
    }
    .to_owned();
    ItemAdjudicationResult {
        context,
        voting_result,
        classification_scope,
        neutral_rescued,
        fileable_oos,
        score_kind: score_kind.clone(),
        score_result: score_result.clone(),
        accepted_weight,
        unique_finder_eligible: score_kind == "finding" && score_result == "accepted",
        ledger_outcome,
        reroute_marker: if neutral_rescued {
            "neutral-rescued".to_owned()
        } else {
            String::new()
        },
        artifact_bucket,
        security: None,
    }
}

/// Execute prepare, serialize, security, then publish in order, stopping at the first error.
///
/// # Errors
///
/// Returns the first hook error without preparing later items.
pub fn run_items<E>(
    contexts: impl IntoIterator<Item = ItemContext>,
    mut serialize: impl FnMut(&ItemAdjudicationResult) -> Result<(), E>,
    mut security_hook: impl FnMut(&ItemContext) -> Result<bool, E>,
    mut publish: impl FnMut(&ItemAdjudicationResult) -> Result<(), E>,
) -> Result<Vec<ItemAdjudicationResult>, E> {
    let mut completed = Vec::new();
    for context in contexts {
        let mut result = adjudicate_item(context);
        serialize(&result)?;
        result.security = Some(security_hook(&result.context)?);
        publish(&result)?;
        completed.push(result);
    }
    Ok(completed)
}
